### PR TITLE
Redirect post URLs with old or wrong titles

### LINKED ISSUE #164: Post URLs with old or incorrect titles are broken
## Problem

Currently, the format for blog post URLs is `/blog/<8-char-id>/post_title_snake_case`. Both segments must match, i.e., a URL with an incorrect post title is a broken URL.

This means if a post has its title edited, then all links containing the old post title immediately become obsolete, resulting in, e.g., broken bookmarks. One funny but unfortunate consequence is that, when the user edits the title of a post and submits, instead of being shown the new content, the user is redirected to the home page from the (now broken) old post link.

## Suggestions

- The title segment of a post URL should reflect the most up-to-date title of the post.
- URLs with correct post IDs but wrong titles (including old titles) should self-correct to show the most up-to-date titles (Stack Exchange does this[^1]).
- (Optional) Refer to the title segment of a URL as the *URL slug*[^2] in subsequent code and discussions.

[^1]: https://meta.stackexchange.com/questions/108949/question-url-slugs-based-on-title
[^2]: https://en.wikipedia.org/wiki/Clean_URL#Slug

### PR DESCRIPTION
**Pull request by flybot-nam-nguyenhoai**
## Closes #164

- [x] Redirect post URLs with wrong or old titles.

    Post URLs such as

    - `/blog/post-id-ending/wrong_title`
    - `/blog/post-id-ending/old_title`
    - `/blog/post-id-ending/` (no title)
    - `/blog/post-id-ending` (no slash, no title)

    are redirected to `/blog/post-id-ending/latest_title`.

### Details

The router dispatches a re-frame event effect to redirect instead of querying the correct post title directly.[^1][^2]

The event is dispatched just before page rendering rather than inside the `:start` route controller. This is due to a race condition—the controller code is sometimes run before the full list of posts is fully loaded, so an event dispatched inside the controller would not be able to query the correct title from the incomplete post list.

[^1]: https://day8.github.io/re-frame/FAQs/UseASubscriptionInAJsEvent/
[^2]: https://day8.github.io/re-frame/FAQs/UseASubscriptionInAnEventHandler/

### CODE REVIEW

**Comment by skydread1** [comment]
There is a race condition between the dispatch of the event `:evt.app/initialize` and the reitit router controller execution. The posts get populated after the controller dispatch the event `:evt.nav/redirect-post-url`.

I think there is no need for controllers (unlike what I initially thought), we can just dispatch `:evt.nav/redirect-post-url` from `blog-single-post-page`:

```clojure
(defn blog-single-post-page
  "Given the blog post identifier, returns the corresponding post in a page.
  If no matching post is found, returns some placeholder content."
  []
  (let [{query-id-ending '?id-ending
         query-url-identifier '?url-identifier}
        @(rf/subscribe [:subs/pattern
                        {:app/current-view
                         {:parameters
                          {:path
                           {:id-ending '?id-ending
                            :url-identifier '?url-identifier}}}}])
        _ (rf/dispatch [:evt.nav/redirect-post-url
                        :flybot/blog-post
                        :blog
                        query-id-ending
                        query-url-identifier])
        queried-post (-> (get-post-by-id-page :blog
                                              query-id-ending
                                              query-url-identifier)
                         vals
                         first
                         add-post-hiccup-content)]
    [...]))
```

File: client/common/src/flybot/client/common/db/event.cljs [hunk 1]
```diff
@@ -14,6 +16,34 @@
...
+     (if (or (nil? queried-post) (matches-url-identifier? queried-post))
+       ;; Don't redirect if there're no page-ID matches, or if there's a
+       ;; complete match
+       {}
+       ;; Redirect on partial match
+       {:fx.router/replace-state
+        [name
+         {:id-ending id-ending
+          :url-identifier (post/post-url-identifier queried-post)}]}))))
+
```
Comment (skydread1) [inline client/common/src/flybot/client/common/db/event.cljs:46]: This logic is only relevant for the `web` frontend so it needs to be moved to `flybot.client.web.core.db.event`

**Review by skydread1 — CHANGES_REQUESTED** [review]

File: client/common/src/flybot/client/common/db/event.cljs (hunk 1 above)
Comment (flybot-nam-nguyenhoai) [reply client/common/src/flybot/client/common/db/event.cljs]: Relocated.

File: client/web/test/flybot/client/web/core/router_test.cljs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                              {:app/current-view
+                               {:data
+                                {:name '?name
+                                 :page-name '?page-name}
+                                :path-params
+                                {:id-ending '?id-ending
+                                 :url-identifier '?url-identifier}}}])
+           go-to-link (fn [path] (->> path
+                                      (r/match-by-path sut/router)
+                                      sut/on-navigate))]
```
Comment (skydread1) [inline client/web/test/flybot/client/web/core/router_test.cljs:61]: Here you can call the dispatch for your post redirect

```clojure
go-to-link (fn [path params] 
                        (rfe/push-state path params)
                        (rf/dispatch [:evt.nav/redirect-post-url
                                      :flybot/blog-post
                                      :blog
                                      {:id-ending params}
                                      {:url-identifier params}]))
```

Then you can call in your test:

```clojure
(testing "Link with wrong title goes to correct view:"
             (go-to-link :flybot/blog-post
                         {:id-ending "-temp-id"
                          :url-identifier "wrong_title"})
             (is (= expected-view (dissoc @get-current-view '&?))))
```

**Review by skydread1 — CHANGES_REQUESTED** [review]

File: client/web/test/flybot/client/web/core/router_test.cljs (hunk 2 above)
Comment (flybot-nam-nguyenhoai) [inline client/web/test/flybot/client/web/core/router_test.cljs:61]: `go-to-link` doesn't seem to work in the REPL.

File: client/web/test/flybot/client/web/core/router_test.cljs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                              '?url-identifier "New_Post_Has_Arrived"}]
+           (testing "Correct link goes to correct view:"
+             (go-to-link "/blog/-temp-id/New_Post_Has_Arrived")
+             (is (= expected-view (dissoc @get-current-view '&?))))
+           (testing "Link with wrong title goes to correct view:"
+             (go-to-link "/blog/-temp-id/wrong_title")
+             (is (= expected-view (dissoc @get-current-view '&?))))
+           (testing "Link with no title goes to correct view:"
+             (go-to-link "/blog/-temp-id/")
+             (is (= expected-view (dissoc @get-current-view '&?))))))
```
Comment (flybot-nam-nguyenhoai) [inline client/web/test/flybot/client/web/core/router_test.cljs:94]: These tests are failing: `@get-current-view` contains the wrong path parameters before the redirect, not the correct parameters after.

File: client/web/test/flybot/client/web/core/router_test.cljs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+             (is (= expected-view (dissoc @get-current-view '&?))))
+           (testing "Link with old title goes to correct view:"
+             (go-to-link "/blog/-temp-id/New_Post_Has_Arrived")
+             (is (= expected-view (dissoc @get-current-view '&?))))
+           (testing "Link with wrong title goes to correct view:"
+             (go-to-link "/blog/-temp-id/wrong_title")
+             (is (= expected-view (dissoc @get-current-view '&?))))
+           (testing "Link with no title goes to correct view:"
+             (go-to-link "/blog/-temp-id/")
+             (is (= expected-view (dissoc @get-current-view '&?))))))))))
```
Comment (flybot-nam-nguyenhoai) [inline client/web/test/flybot/client/web/core/router_test.cljs:128]: These tests are failing: @get-current-view contains the wrong path parameters before the redirect, not the correct parameters after.

File: client/web/test/flybot/client/web/core/router_test.cljs (hunk 2 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/test/flybot/client/web/core/router_test.cljs:61]: I think I did yesterday. Let me try again.

Comment (flybot-nam-nguyenhoai) [reply client/web/test/flybot/client/web/core/router_test.cljs]: It works, thank you!

File: client/common/src/flybot/client/common/db/fx.cljs [hunk 5]
```diff
@@ -1,5 +1,18 @@
 (ns flybot.client.common.db.fx
-  (:require [re-frame.core :as rf]))
+  (:require [re-frame.core :as rf]
+            [reitit.frontend.easy :as rfe]))
+
+;; -- Producing pure results ---
+
+(rf/reg-fx
+ :fx.pure/pure
+ identity)
```
Comment (skydread1) [inline client/common/src/flybot/client/common/db/fx.cljs:9]: To be removed

File: client/web/src/flybot/client/web/core/db/event.cljs [hunk 6]
```diff
@@ -1,10 +1,14 @@
 (ns flybot.client.web.core.db.event
-  (:require [flybot.client.common.db.event]
-            [flybot.common.utils :as utils :refer [toggle]]
-            [ajax.edn :refer [edn-request-format edn-response-format]]
+  (:require [ajax.edn :refer [edn-request-format edn-response-format]]
+            [clojure.string :as str]
             [day8.re-frame.http-fx]
+            [flybot.client.common.db.event]
+            [flybot.client.web.core.dom.page.post :as post]
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/db/event.cljs:6]: We do not want deps to the `dom` here.

`post-url-identifier` should be moved in a more relevant ns.

File: client/web/test/flybot/client/web/core/router_test.cljs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                        (let [[name path-params :as match]
+                              ((pull/qfn
+                                {:data
+                                 {:name '?name}
+                                 :path-params '?path-params}
+                                [?name ?path-params])
+                               (r/match-by-path sut/router path))]
+                          (rfe/replace-state name path-params)
+                          (rf/dispatch [:evt.nav/redirect-post-url])
+                          (print match)))]
```
Comment (skydread1) [inline client/web/test/flybot/client/web/core/router_test.cljs:71]: `print` to be removed

File: client/web/src/flybot/client/web/core/db/event.cljs [hunk 8]
```diff
@@ -18,6 +22,46 @@
...
+                       first)]
+     (if (or (nil? queried-post) (matches-url-identifier? queried-post))
+       ;; Don't redirect if there're no page-ID matches, or if there's a
+       ;; complete match
+       {}
+       ;; Redirect on partial match
+       {:fx.router/replace-state
+        [(get router/redirect-name-map name)
+         {:id-ending id-ending
+          :url-identifier (post/post-url-identifier queried-post)}]}))))
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/db/event.cljs:63]: More compact impl:

```clojure
(when-not (or (not queried-post) (matches-url-identifier? queried-post))
       ;; Redirect on partial match
       {:fx.router/replace-state
        [(get router/redirect-name-map name)
         {:id-ending id-ending
          :url-identifier (post/post-url-identifier queried-post)}]})
```

**Review by skydread1 — CHANGES_REQUESTED** [review]

File: client/web/src/flybot/client/web/core/router.cljs [hunk 9]
```diff
@@ -29,14 +29,26 @@
...
+    {:page-name :blog
+     :view #(do
+              (rf/dispatch [:evt.nav/redirect-post-url])
+              (blog-single-post-page))}
+    [""
+     {:name :flybot/-blog>id}]
+    ["/"
+     {:name :flybot/-blog>id>}]
+    ["/:url-identifier"
+     {:name :flybot/blog-post}]]
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/router.cljs:42]: Good!

I suggest to use a dedicated key for the redirect instead of a mapping such as:

```clojure
["/blog/:id-ending"
    {:page-name :blog
     :view #(do
              (rf/dispatch [:evt.nav/redirect-post-url])
              (blog-single-post-page))}
    [""
     {:redirect-name :flybot/blog-post}]
    ["/"
     {:redirect-name :flybot/blog-post}]
    ["/:url-identifier"
     {:name :flybot/blog-post}]]
```

Then you won't trigger an error for same name route in reitit and you can easily access it in your event `evt.nav/redirect-post-url`:

```clojure
{:fx.router/replace-state
        [(or name redirect-name)
         ...]}
```

**Review by skydread1 — CHANGES_REQUESTED** [review]

File: client/web/src/flybot/client/web/core/db/event.cljs (hunk 6 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/db/event.cljs:6]: This will be tricky. `post-url-identifier` uses a post's Hiccup content to find its title.

File: client/web/src/flybot/client/web/core/router.cljs (hunk 9 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/router.cljs:42]: I thought of it, but I need both redirected routes to also be named so that I can actually go to these routes using `replace-state` in my router tests.

This also makes `(or name redirect-name)` not work.

File: client/web/test/flybot/client/web/core/router_test.cljs (hunk 7 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/test/flybot/client/web/core/router_test.cljs]: Removed.

File: client/web/src/flybot/client/web/core/db/event.cljs (hunk 8 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/db/event.cljs]: Fixed.

```clojure
(when (and queried-post (not (matches-url-identifier? queried-post))) …
```
to avoid double negatives.

File: client/web/src/flybot/client/web/core/router.cljs (hunk 9 above)
Comment (skydread1) [reply client/web/src/flybot/client/web/core/router.cljs:42]:
```clojure
(rf/reg-event-fx
 :evt.nav/redirect-post-url
 (fn [{:keys [db]} [_]]
   (let [[name redirect-name page-name id-ending url-identifier]
         ((pull/qfn {:app/current-view
                     {:data
                      {:name '?name
                       :redirect-name '?redirect-name
                       :page-name '?page-name}
                      :path-params
                      {:id-ending '?id-ending
                       :url-identifier '?url-identifier}}}
                    [?name ?redirect-name ?page-name ?id-ending ?url-identifier])
          db)
         matches-page? (fn [post] (-> post :post/page (= page-name)))
         matches-id-ending? (fn [id] (str/ends-with? (str id) id-ending))
         matches-url-identifier? (fn [post]
                                   (= url-identifier
                                      (post/post-url-identifier post)))
         queried-post (->>
                       (:app/posts db)
                       (into {}
                             (filter (fn [[id post]]
                                       (and (matches-id-ending? id)
                                            (matches-page? post)))))
                       vals
                       first)]
     (when (and queried-post (not (matches-url-identifier? queried-post)))
       ;; Redirect on partial match (when everything matches except post title)
       {:fx.router/replace-state
        [(or name redirect-name)
         {:id-ending id-ending
          :url-identifier (post/post-url-identifier queried-post)}]}))))
```

File: client/web/src/flybot/client/web/core/db/event.cljs [hunk 10]
```diff
@@ -1,10 +1,14 @@
 (ns flybot.client.web.core.db.event
-  (:require [flybot.client.common.db.event]
-            [flybot.common.utils :as utils :refer [toggle]]
-            [ajax.edn :refer [edn-request-format edn-response-format]]
+  (:require [ajax.edn :refer [edn-request-format edn-response-format]]
+            [clojure.string :as str]
             [day8.re-frame.http-fx]
+            [flybot.client.common.db.event]
+            [flybot.client.web.core.dom.page.post :as post]
+            [flybot.client.web.core.router :as router]
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/db/event.cljs:7]: no need this deps anymore

File: client/web/test/flybot/client/web/core/router_test.cljs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  (:require
+   [cljs.test :refer [deftest is testing use-fixtures]]
+   [day8.re-frame.test :as rf-test]
+   [flybot.client.web.core.router :as sut]
+   [flybot.common.test-sample-data :as s]
+   [flybot.common.utils :as utils]
+   [re-frame.core :as rf]
+   [reitit.core :as r]
+   [reitit.frontend.easy :as rfe]
+   [sg.flybot.pullable :as pull]))
```
Comment (skydread1) [inline client/web/test/flybot/client/web/core/router_test.cljs:11]: no need these deps anymore

**Review by skydread1 — CHANGES_REQUESTED** [review]
Good job!

File: client/web/src/flybot/client/web/core/db/event.cljs (hunk 10 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/db/event.cljs]: Just `router`, right? I believe we still need `db.event` and `post`.

File: client/web/test/flybot/client/web/core/router_test.cljs (hunk 11 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/test/flybot/client/web/core/router_test.cljs]: Removed.

File: client/web/src/flybot/client/web/core/router.cljs (hunk 9 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/router.cljs]: Switched to using `:redirect-name` and rewrote tests.

**Review by skydread1 — APPROVED** [review]
